crypto/crypto.c

```diff
@@ -0,0 +1,425 @@
+    char *header = NULL, *name = NULL;
+    //returns 0 for fail and 1 on success
+    rc = !PEM_read_bio(bio, &name, &header, output, (long int *)olen);
+    if (header) free(header);
+    if (name) free(name);
+    BIO_free(bio);
+    return rc;
+}
+
+/*=====================END OPENSSL FUNCTIONS=====================*/
+#else
+/*====================START MBEDTLS FUNCTIONS====================*/
+
+
+
+void *crypto_pkcs7_parse_der(const unsigned char *buf, const int buflen) 
+{
+    int rc;
+    struct mbedtls_pkcs7 *pkcs7;
+    pkcs7 = malloc(sizeof(struct mbedtls_pkcs7));
+    if (!pkcs7) {
+     prlog(PR_ERR, "ERROR: failed to allocate memory\n");
+     return NULL;
+    }
+    mbedtls_pkcs7_init(pkcs7);
+    rc = mbedtls_pkcs7_parse_der(buf, buflen, pkcs7);
+    if (rc != MBEDTLS_PKCS7_SIGNED_DATA)  // if pkcs7 parsing fails, then try new signed data format 
+            prlog(PR_ERR, "ERROR: parsing pkcs7 failed mbedtls error #%04x\n", rc); 
+    else 
+        rc = SUCCESS;
+
+    if (rc) {
+        crypto_pkcs7_free(pkcs7);
+        return NULL;
+    }
+    else
+        return (void *)pkcs7;
+}
+
+// size_t crypto_get_pkcs7_struct_size() 
+// {
+//     return sizeof(struct mbedtls_pkcs7);
+// }
+
+int crypto_pkcs7_is_md_sha256(void *pkcs7)
+{
+    return memcmp(((struct mbedtls_pkcs7 *)pkcs7)->signed_data.digest_alg_identifiers.p, MBEDTLS_OID_DIGEST_ALG_SHA256, strlen(MBEDTLS_OID_DIGEST_ALG_SHA256) ) ;
+}
+void crypto_pkcs7_free(void *pkcs7)
+{
+        mbedtls_pkcs7_free((struct mbedtls_pkcs7 *)pkcs7);
+        free(pkcs7);
+}
+
+void* crypto_get_signing_cert(void *pkcs7, int cert_num)
+{
+    mbedtls_x509_crt *pkcs7_cert = NULL;
+
+    pkcs7_cert = &((struct mbedtls_pkcs7 *)pkcs7)->signed_data.certs;
+    for (int i = 0; i < cert_num && pkcs7_cert != NULL; i++)
+        pkcs7_cert = pkcs7_cert->next;
+
+    return (void *)pkcs7_cert;
+
+}
+
+int crypto_get_x509_der_len(void *x509) 
+{
+    return ((mbedtls_x509_crt *)x509)->raw.len; 
+}
+
+int crypto_get_tbs_x509_der_len(void *x509) 
+{
+    return ((mbedtls_x509_crt *)x509)->tbs.len;
+}
+
+int crypto_get_x509_version(void *x509) 
+{
+    return ((mbedtls_x509_crt *)x509)->version;
+}
+
+int crypto_x509_is_RSA(void *x509)
+{
+    int rc;
+    rc = ((mbedtls_x509_crt *)x509)->pk.pk_info->type;
+    if (rc != MBEDTLS_PK_RSA)
+        //zero is also a pk type (MBEDTLS_PK_NONE) so return generic failure if zero so it doesnt look like a success
+       return  (rc == 0 ? CERT_FAIL : rc);
+    else 
+        return SUCCESS;
+}
+
+int crypto_get_x509_sig_len(void *x509)
+{
+    return ((mbedtls_x509_crt *)x509)->sig.len;
+}
+
+int crypto_x509_md_is_sha256(void *x509)
+{
+    if (((mbedtls_x509_crt *)x509)->sig_md == MBEDTLS_MD_SHA256) 
+        return SUCCESS;
+    else
+        return CERT_FAIL;
+}
+
+int crypto_x509_oid_is_pkcs1_sha256(void *x509)
+{
+    if (strncmp((const char *)((mbedtls_x509_crt *)x509)->sig_oid.p, MBEDTLS_OID_PKCS1_SHA256, ((mbedtls_x509_crt *)x509)->sig_oid.len) )
```

Comment:
MBEDTLS_OID_CMP

Reply:
Here, if the sig_oid.len > MBEDTLS_OID_PKCS1_SHA256 len, you'll read off into random global storage space.